pkg/operator/defaultscccontroller/defaultscc.go

```diff
@@ -0,0 +1,98 @@
+package defaultscccontroller
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	"github.com/openshift/api"
+	securityv1 "github.com/openshift/api/security/v1"
+
+	assets "github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/defaultscc_assets"
+)
+
+// DefaultSCCCache maintains the list of default SecurityContextConstraints objects
+// shipped wth an OpenShift cluster.
+type DefaultSCCCache struct {
+	orderedNames []string
+	set          map[string]*securityv1.SecurityContextConstraints
+}
+
+// Get returns the SecurityContextConstraints object associated with the
+// specified name. The returned SecurityContextConstraints is a deep copy and so
+// it can be modified at will without impacting the integrity of the underlying cache.
+func (d *DefaultSCCCache) Get(name string) (scc *securityv1.SecurityContextConstraints, exists bool) {
+	obj, exists := d.set[name]
+	if !exists {
+		return
+	}
+
+	scc = obj.DeepCopy()
+	return
+}
+
+func (d *DefaultSCCCache) DefaultSCCNames() []string {
+	return d.orderedNames
+}
+
+// NewDefaultSCCCache renders the assets associated with the default set of SCC and
+// loads it into a cache (DefaultSCCCache)
+func NewDefaultSCCCache() (cache *DefaultSCCCache, err error) {
+	decoder, decoderErr := decoder()
+	if decoderErr != nil {
+		err = fmt.Errorf("failed to create decoder - %s", decoderErr.Error())
+		return
+	}
+
+	names := make([]string, 0)
+	set := make(map[string]*securityv1.SecurityContextConstraints)
+	for _, name := range assets.AssetNames() {
```

Comment:
are asset names ordered? Can a new and old operator with different bindata race?

Reply:
we don't depend on the order in which `assets.AssetNames()` returns the names. The objects are stored in a map and since iterating a map does not guarantee any particular order, we store the names in a slice and use it to iterate.

The reason we want a consistent order (be it any) is so that the error message (that ends up in status. conditions) `default scc has been mutated [anyuid hostaccess]` is consistent as well.

Reply:
I thought the asset name order determines the order in those error messages. If this is not constant, condition will change just because the operator has been updated. Not super critical, yet little ugly.

Reply:
yes, asset name order determines the order of the scc name in the error message. `assets.AssetNames()` returns from a map and thus the order is indeterminate. I am now sorting the `names` slice. Also, I have added a unit test test verifies that the order is fixed. So, we cover the following cases:
- If the operator process restarts we will see the same ordering now.
- If we make change to the code and the order is violated, the unit test will fail to alert us.